### PR TITLE
[enhancement] UL TFT support in S1 SIM

### PR DESCRIPTION
**Pull request by pruthvihebbani**
## Title
Added UL TFT support in S1 SIM

## Summary
Added code to store the UL TFT packet filter/s in the order of precedence and send the UL IP packet on the dedicated bearer if the packet filter matches. If not the packet is sent on default bearer.

## Test plan
- Executed test_attach_detach_rar_tcp_data.py test case from PR #1442

### CODE REVIEW

File: TestCntlrApp/src/enbApp/nb_dam.c [hunk 1]
```diff
@@ -481,6 +658,7 @@ NbDamTnlInfo                 *tnlInfo
    NbDamTnlType    tnlType = tnlInfo->tnlType;
    NbDamDrbCb      *drbCb  = NULLP;
    NbIpInfo        *ipInfo = NULLP;
+   U8 idx = 0;
```
Comment (ulaskozat) [inline TestCntlrApp/src/enbApp/nb_dam.c:661]: move inside the for loop.

File: TestCntlrApp/src/enbApp/nb_dam.c [hunk 2]
```diff
@@ -705,65 +888,122 @@ U8                           msgType
...
+
+  SFndLenMsg(mBuf, &len);
+  if (len == 0) {
+    NB_LOG_ERROR(&nbCb, "Empty Buffer recieved");
+    SPutMsg(mBuf);
+    RETVALUE(ROK);
+  }
+
+  /* Fetch ToS or DSCP*/
+  for (idx = 0; idx < 1; idx++) {
```
Comment (ulaskozat) [inline TestCntlrApp/src/enbApp/nb_dam.c:914]: why for loop if iterating only once?

File: TestCntlrApp/src/enbApp/nb_dam.c [hunk 3]
```diff
@@ -705,65 +888,122 @@ U8                           msgType
...
-   }
-
-   /* mark for ue data received */
-   ueCb->dataRcvd = TRUE;   
-   /* get the tunnel Cb */
-   if ( ROK != (cmHashListFind(&(ueCb->drbs), (U8 *)&(drbId),
-               sizeof(U8), 0, (PTR *)&drbCb)))
-   {    
+    }
+    ipIdx++;
```
Comment (ulaskozat) [inline TestCntlrApp/src/enbApp/nb_dam.c:920]: not needed when for loop is gone.

File: TestCntlrApp/src/enbApp/nb_dam.c [hunk 4]
```diff
@@ -705,65 +888,122 @@ U8                           msgType
...
+  }
+  /* Tos or DSCP*/
+  ipPktFields.srvClass = ipPkt[0];
+  /* Skip 9 bytes to fetch the protocol ID*/
+  ipIdx = 9;
+  /* Fetch protocol Id*/
+  for (idx = 0; idx < 1; idx++) {
```
Comment (ulaskozat) [inline TestCntlrApp/src/enbApp/nb_dam.c:927]: again why for loop if iterated only once?

File: TestCntlrApp/src/enbApp/nb_dam.c [hunk 5]
```diff
@@ -705,65 +888,122 @@ U8                           msgType
...
+      NB_LOG_ERROR(&nbCb, "Failed to fetch protocol Id");
+      SPutMsg(mBuf);
       RETVALUE(RFAILED);
-   }    
-   else 
-   {    
-      tnl = drbCb->tnlInfo;   
-   }    
+    }
+    ipIdx++;
```
Comment (ulaskozat) [inline TestCntlrApp/src/enbApp/nb_dam.c:933]: no need without for loop...

File: TestCntlrApp/src/enbApp/nb_dam.c [hunk 6]
```diff
@@ -1516,6 +1797,80 @@ PRIVATE S16 nbDamBndLSap
    RETVALUE (ROK);
 } /* nbBndLSap */
 
+/** @brief This function matches the Packet filters with the ip fileds of the
```
Comment (ulaskozat) [inline TestCntlrApp/src/enbApp/nb_dam.c:1800]: typo: fileds->fields

File: TestCntlrApp/src/enbApp/nb_dam.c [hunk 7]
```diff
@@ -911,26 +908,21 @@ PUBLIC S16 nbDamPcapDatInd(Buffer *mBuf)
...
   /* Fetch ToS or DSCP*/
-  for (idx = 0; idx < 1; idx++) {
-    if ((SExamMsg(&ipPkt[idx], mBuf, ipIdx) != ROK)) {
-      NB_LOG_ERROR(&nbCb, "Failed to fetch ToS");
-      SPutMsg(mBuf);
-      RETVALUE(RFAILED);
-    }
-    ipIdx++;
+  U8 idx = 0;
+  if ((SExamMsg(&ipPkt[idx], mBuf, ipIdx) != ROK)) {
```
Comment (ulaskozat) [inline TestCntlrApp/src/enbApp/nb_dam.c:912]: why not directly write `&ipPkt[0]` instead of `&ipPkt[idx]`

File: TestCntlrApp/src/enbApp/nb_dam.c [hunk 8]
```diff
@@ -911,26 +908,21 @@ PUBLIC S16 nbDamPcapDatInd(Buffer *mBuf)
...
   ipIdx = 9;
   /* Fetch protocol Id*/
-  for (idx = 0; idx < 1; idx++) {
-    if ((SExamMsg(&ipPkt[idx], mBuf, ipIdx) != ROK)) {
-      NB_LOG_ERROR(&nbCb, "Failed to fetch protocol Id");
-      SPutMsg(mBuf);
-      RETVALUE(RFAILED);
-    }
-    ipIdx++;
+  if ((SExamMsg(&ipPkt[idx], mBuf, ipIdx) != ROK)) {
```
Comment (ulaskozat) [inline TestCntlrApp/src/enbApp/nb_dam.c:922]: same here, you can directly use `&ipPkt[0]`

Comment (pruthvihebbani) [reply TestCntlrApp/src/enbApp/nb_dam.c]: done

File: TestCntlrApp/src/enbApp/nb_dam.c (hunk 7 above)
Comment (pruthvihebbani) [reply TestCntlrApp/src/enbApp/nb_dam.c]: done

File: TestCntlrApp/src/enbApp/nb_dam.c (hunk 6 above)
Comment (pruthvihebbani) [reply TestCntlrApp/src/enbApp/nb_dam.c]: done

File: TestCntlrApp/src/enbApp/nb_dam.c (hunk 5 above)
Comment (pruthvihebbani) [reply TestCntlrApp/src/enbApp/nb_dam.c]: done

File: TestCntlrApp/src/enbApp/nb_dam.c (hunk 4 above)
Comment (pruthvihebbani) [reply TestCntlrApp/src/enbApp/nb_dam.c]: done

File: TestCntlrApp/src/enbApp/nb_dam.c (hunk 1 above)
Comment (pruthvihebbani) [reply TestCntlrApp/src/enbApp/nb_dam.c]: done

File: TestCntlrApp/src/enbApp/nb_dam.c (hunk 2 above)
Comment (pruthvihebbani) [reply TestCntlrApp/src/enbApp/nb_dam.c]: done

File: TestCntlrApp/src/enbApp/nb_dam.c (hunk 3 above)
Comment (pruthvihebbani) [reply TestCntlrApp/src/enbApp/nb_dam.c]: done

**Pull request merged**